### PR TITLE
plugin/forward: add query timeout

### PR DESCRIPTION
**Pull request by rdrozhdzh**
- subsequently retry sending query via different proxies
   until query timeout occurs. A proxy can be tried again if
   all other proxies were already tried

this fix is an alternative to #1624

### CODE REVIEW

**Comment by miekg** [comment]
yes, I think this is a saner way of doing things. I'll take a closer look later (on holiday atm)

File: plugin/forward/forward.go [hunk 1]
```diff
@@ -65,24 +66,36 @@ func (f *Forward) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg
 	}
 
 	fails := 0
-	var span, child ot.Span
-	var upstreamErr error
-	span = ot.SpanFromContext(ctx)
+	i := 0
+	returnErr := errNoHealthy
```
Comment (miekg) [inline plugin/forward/forward.go:70]: I thought a bit about the name `upstreamErr` and I think it's a better name than `returnErr`

File: plugin/forward/forward.go [hunk 2]
```diff
@@ -185,3 +189,5 @@ const (
 	randomPolicy policy = iota
 	roundRobinPolicy
 )
+
+const queryTimeout = 5 * time.Second
```
Comment (miekg) [inline plugin/forward/forward.go:193]: `defaultTimeout` ?

File: plugin/forward/forward.go [hunk 3]
```diff
@@ -65,24 +66,36 @@ func (f *Forward) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg
...
+	span := ot.SpanFromContext(ctx)
+	proxyList := f.list()
+	deadline := time.Now().Add(queryTimeout)
+
+	for time.Now().Before(deadline) {
+		if i >= len(proxyList) {
+			if i == fails {
+				// All upstream proxies are dead, assume healtcheck is completely broken and randomly
+				// select an upstream to connect to.
+				i = rand.Intn(len(proxyList))
```
Comment (miekg) [inline plugin/forward/forward.go:80]: Just leave the original code here?

File: plugin/forward/forward.go [hunk 4]
```diff
@@ -65,24 +66,36 @@ func (f *Forward) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg
...
+				deadline = time.Now() // give it a last attempt
+
+				HealthcheckBrokenCount.Add(1)
```
Comment (miekg) [inline plugin/forward/forward.go:83]: add a `continue` here, so we can drop the else and dedent the body of the else.

File: plugin/forward/forward.go [hunk 5]
```diff
@@ -65,24 +66,36 @@ func (f *Forward) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg
 	}
 
 	fails := 0
-	var span, child ot.Span
-	var upstreamErr error
-	span = ot.SpanFromContext(ctx)
+	i := 0
+	returnErr := errNoHealthy
+	span := ot.SpanFromContext(ctx)
```
Comment (miekg) [inline plugin/forward/forward.go:71]: why did you move the `child` to down below?

File: plugin/forward/forward.go [hunk 6]
```diff
@@ -65,24 +66,36 @@ func (f *Forward) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg
...
+	proxyList := f.list()
```
Comment (miekg) [inline plugin/forward/forward.go:72]: just use `list`

File: plugin/forward/health_test.go [hunk 7]
```diff
@@ -45,15 +45,21 @@ func TestHealth(t *testing.T) {
...
 	const expected = 1
 	i := uint32(0)
+	q := uint32(0)
 	s := dnstest.NewServer(func(w dns.ResponseWriter, r *dns.Msg) {
 		if r.Question[0].Name == "." {
 			// health check, answer
 			atomic.AddUint32(&i, 1)
 			ret := new(dns.Msg)
 			ret.SetReply(r)
 			w.WriteMsg(ret)
```
Comment (miekg) [inline plugin/forward/health_test.go:55]: just `return` here

File: plugin/pkg/up/up.go [hunk 8]
```diff
@@ -37,9 +38,11 @@ func (p *Probe) Stop() { p.stop <- true }
 func (p *Probe) Start(interval time.Duration) { go p.start(interval) }
 
 func (p *Probe) start(interval time.Duration) {
+	stop := uint32(0)
 	for {
 		select {
 		case <-p.stop:
+			atomic.StoreUint32(&stop, 1)
```
Comment (miekg) [inline plugin/pkg/up/up.go:45]: this is purely for testing?

**Review by miekg — CHANGES_REQUESTED** [review]
Generally OK with this change; bunch of comments, some nits, some not.

File: plugin/forward/forward.go (hunk 1 above)
Comment (rdrozhdzh) [reply plugin/forward/forward.go:70]: My intention was to use the same var for upstream error and no healthy error, and get rid of the extra 'if' statement at the end of ServeDNS(). Therefore, I renamed it

File: plugin/forward/forward.go (hunk 2 above)
Comment (rdrozhdzh) [reply plugin/forward/forward.go:193]: I can change to defaultTimeout if you wish. But this name tell us less about purpose of the constant

File: plugin/forward/forward.go (hunk 3 above)
Comment (rdrozhdzh) [reply plugin/forward/forward.go:80]: the original code was more expensive from performance point of view, so I changed it

File: plugin/forward/forward.go (hunk 4 above)
Comment (rdrozhdzh) [reply plugin/forward/forward.go:83]: Well, actually here we detected that all proxies are down, and we need to do the last attempt to connect to random proxy. This logic was in original code. So, I calculate index of random proxy in line 80, specify last attempt by resetting `deadline`, and go to the main part of loop where proxy is taken from list by index and `Proxy.connect()` is called.
Adding `continue` here will skip connecting to proxy

File: plugin/forward/forward.go (hunk 5 above)
Comment (rdrozhdzh) [reply plugin/forward/forward.go:71]: 1. To reduce the scope of `child`. It is not referred outside the loop
2. This way `child` is initialized with `nil` in each iteration. Theoretically `child` may not be assigned with value in current iteration, and later down the code we compare it with `nil` to call `Finish()`. Non-nil value of `child` from previous iteration could lead to repeated call to `Finish()` which seems incorrect. Though, I guess such case is unlikely on practice

File: plugin/pkg/up/up.go (hunk 8 above)
Comment (rdrozhdzh) [reply plugin/pkg/up/up.go:45]: Actually not. The test run successfully without this change. The goroutine below is stopped at the end of test and doesn't affect results
This is rather a fix of probe logic. If we call `Probe.Stop()` we expect it will stop probing. But before my fix in case if probing function always return error it keeps probing forever, which seems incorrect

**Comment by rdrozhdzh** [comment]
Thank you @miekg for review. It seems I did too much refactoring in this PR :) I can revert the changes not related directly to the topic of this PR. Just let me know if you don't want some changes to be merged here

**Comment by miekg** [comment]
Quick reply: yes minimizing the changes helps to get a pr approved. If you
want subsequent changes just submit further PRs

**Comment by rdrozhdzh** [comment]
The new PR#1665 created as a replacement for this one, so closing this
